persist/fs/mmap_linux.go

```diff
@@ -56,22 +56,42 @@ func mmap(fd, offset, length int64, flags int, opts mmapOptions) ([]byte, error)
 		prot = prot | syscall.PROT_WRITE
 	}
 
-	b, err := syscall.Mmap(int(fd), offset, int(length), prot, flags)
-	if err != nil {
-		return nil, fmt.Errorf("mmap error: %v", err)
+	flagsWithoutHugeTLB := flags
+	shouldUseHugeTLB := opts.hugeTLB.enabled && length >= opts.hugeTLB.threshold
+	if shouldUseHugeTLB {
+		// We use the MAP_HUGETLB flag instead of MADV_HUGEPAGE because transparent
+		// hugepages only work with anonymous, private pages. Please see the MADV_HUGEPAGE
+		// section of http://man7.org/linux/man-pages/man2/madvise.2.html and the MAP_HUGETLB
+		// section of http://man7.org/linux/man-pages/man2/mmap.2.html for more details.
+		flags = flags | syscall.MAP_HUGETLB
 	}
 
-	if !opts.hugePages.enabled ||
-		length < opts.hugePages.threshold {
-		return b, nil
+	b, err := syscall.Mmap(int(fd), offset, int(length), prot, flags)
+	// Sometimes allocations that specify huge pages will fail because the O.S
+	// isn't configured properly or there are not enough available huge pages in
+	// the pool. You can try and allocate more by executing:
+	// 		echo 20 > /proc/sys/vm/nr_hugepages
+	// See this document for more details: https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt
+	// Regardless, we don't want to fail hard in that scenario. Instead, we try
+	// and mmap without the hugeTLB flag.
+	if err != nil && shouldUseHugeTLB {
+		b, err = syscall.Mmap(int(fd), offset, int(length), prot, flagsWithoutHugeTLB)
+		// If it still failed, then return an error
+		if err != nil {
+			return mmapResult{}, fmt.Errorf("mmap error: %v", err)
+		}
+
+		// Otherwise, return success but with an appropriate warning that we were
+		// not able to enable the HUGETLB flag.
+		// TODO:
+		return mmapResult{result: b}, nil
```

Comment:
You can get rid of line 80 - 87 here as the same code sits below.

i.e.
```go
var (b []byte
if err != nil && shouldUseHugeTLB {
  b, err = syscall.Mmap(int(fd), offset, int(length), prot, flagsWithoutHugeTLB)
}

```

Reply:
oh duh, thanks!